### PR TITLE
Improvements to the headers_xss module

### LINKED ISSUE #239: [Medium] Multiple logical bugs within header_xss module
The `header_xss` module, contains the following on line 34:
```
        r = requests.head(host)
        for header in r.headers:
            headers_xss[header] = payloads_xss
        req = requests.post(host, headers=headers_xss)
        if payloads_xss.lower() in req.text.lower():
            return True
        else:
            return False
```
So what the engine does is, it makes a HEAD request to the site, and then it iterates over the received headers from the HEAD request and adds them to the already existing `headers_xss` dictionary. By doing this we are actually violating the w3 standards defined in RFC 7230-7235. 

A quick glance at the [wikipedia page](https://en.wikipedia.org/wiki/List_of_HTTP_header_fields#Request_fields) says a lot about the specific set of header fields which can be used during a client request. But, when we are appending the set of response header fields, to the request header set, we are defying the standard.

Elaborately explaining, taking the following example:
- Site: http://www.webscantest.com
- Making the HEAD request, yeilds the following headers:
```
{
'Date': 'Wed, 25 Mar 2020 06:05:32 GMT', 
'Server': 'Apache/2.4.7 (Ubuntu)', 
'X-Powered-By': 'PHP/5.5.9-1ubuntu4.29', 
'Set-Cookie': 'TEST_SESSIONID=ak6b8793njuaai9lns68ka9q54; path=/, NB_SRVID=srv140717; path=/', 
'Expires': 'Thu, 19 Nov 1981 08:52:00 GMT', 
'Cache-Control': 'no-store, no-cache, must-revalidate, post-check=0, pre-check=0', 
'Pragma': 'no-cache', 
'Connection': 'close', 
'Content-Type': 'text/html'
}
```
- Performing the dict transformation: `for header in r.headers: headers_xss[header] = payloads_xss`, the header set we get is:
```
{
'User-Agent': '<script>alert()</script>', 
'Except': '<script>alert()</script>',
'Accept-Encoding': '<script>alert()</script>', 
'Referer': '<script>alert()</script>', 
'Accept-Language': '<script>alert()</script>', 
'Date': '<script>alert()</script>', 
'Server': '<script>alert()</script>', 
'X-Powered-By': '<script>alert()</script>', 
'Set-Cookie': '<script>alert()</script>', 
'Expires': '<script>alert()</script>', 
'Cache-Control': '<script>alert()</script>', 
'Pragma': '<script>alert()</script>', 
'Connection': '<script>alert()</script>', 
'Content-Type': '<script>alert()</script>'
}
```
Now if we are performing the POST request using this set of headers, we are violating the client-server model, since we are both using the header set of a server as well as the browser/client lol.

Also, I see we are making a POST request without any data?

Also, the `payloads_xss` value is set to `<script>alert(/1/)</script>` which is a HTML context payload. I think we should support javascript context based payloads as well as reflected input based contexts too, since we are never sure how and where the input is getting reflected/stored!

I maybe wrong in my understanding hence ccing @pradeepjairamani, the author of the code. He might have other intentions while writing the code, hence I'm open to a quality discussion. :)

### PR DESCRIPTION
**Pull request by 0xInfection**
#### Checklist
- [x] I have followed the [Contributor Guidelines](https://github.com/zdresearch/OWASP-Nettacker/wiki/Developers#contribution-guidelines).
- [x] I have added the relevant documentation.
- [x] My branch is up-to-date with the Upstream master branch.

#### Changes proposed in this pull request
- [x] Fixes #239 logical flaws.
- [x] Adds more power to the module by adding support for more HTTP methods.

#### Your development environment
- OS: `Ubuntu`
- OS Version: `18.04`
- Python Version: `3.6`

### CODE REVIEW

File: lib/payload/scanner/header_xss/engine.py [hunk 1]
```diff
@@ -13,26 +13,34 @@ def header_xss(host, payloads_xss = None, headers_xss = None):
...
-        r = requests.head(host)
-        for header in r.headers:
-            headers_xss[header] = payloads_xss
-        req = requests.post(host, headers=headers_xss)
-        if payloads_xss.lower() in req.text.lower():
-            return True
-        else:
-            return False
+        # Expanding our scope of attack by using all the
+        # HTTP methods available, HEAD and OPTIONS return
```
Comment (Ali-Razmjoo) [inline lib/payload/scanner/header_xss/engine.py]: Codacy Issue found: [Trailing whitespace](https://app.codacy.com/manual/zdresearch/OWASP-Nettacker/pullRequest?prid=5251812)

**Comment by Ali-Razmjoo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/payload/scanner/header_xss/engine.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zdresearch/OWASP-Nettacker/pullRequest?prid=5251812&bid=17209591)

File: lib/payload/scanner/header_xss/engine.py (hunk 1 above)
Comment (0xInfection) [reply lib/payload/scanner/header_xss/engine.py]: Hey @Ali-Razmjoo, thanks for the notification! Fixed the trailing whitespace.

**Comment by Ali-Razmjoo** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- lib/payload/scanner/header_xss/engine.py  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/zdresearch/OWASP-Nettacker/pullRequest?prid=5251812&bid=17209591)

**Comment by 0xInfection** [comment]
I don't get it. Why is the build failing? @Ali-Razmjoo can you give me an idea whats wrong?

**Comment by the-Blue-Falcon** [comment]
I have used Codacy lately, but their _code quality checker app_ seems to have broken lately. I think this PR is mergeable. :)

Nice work @0xInfection.

File: lib/payload/scanner/header_xss/engine.py [hunk 2]
```diff
@@ -4,7 +4,7 @@
 
 import requests
 
-def header_xss(host, payloads_xss = None, headers_xss = None):
+def header_xss(host, data=None, payloads_xss=None, headers_xss=None):
```
Comment (the-Blue-Falcon) [inline lib/payload/scanner/header_xss/engine.py]: Followed PEP-8 standards here.

**Comment by Ali-Razmjoo** [comment]
thanks for the contribution 👍

**Pull request merged**